main.cpp

```diff
@@ -272,158 +220,95 @@ struct HeapA
 #include <functional>
 #include <memory>
 
-struct DoubleType;
-struct IntType;
-
-struct FloatType
+template<typename T>
+struct Numeric
 {
-    explicit FloatType(float lhs) : value(new float(lhs)) {}
+    using Type = T;
 
-    ~FloatType()
-    {
-        delete value;
-    }
+    explicit Numeric(Type lhs) : value(std::make_unique<Type>(lhs)) {}
 
-    FloatType& operator+=(float rhs);
-    FloatType& operator-=(float rhs);
-    FloatType& operator*=(float rhs);
-    FloatType& operator/=(float rhs);
+    Numeric<Type>& operator+=(Type rhs);
+    Numeric<Type>& operator-=(Type rhs);
+    Numeric<Type>& operator*=(Type rhs);
 
-    operator float() const
+    template<typename Param>
+    Numeric& operator/=(Param rhs )
```

Comment:
pass by const ref to avoid making unnecessary copies.
since `Type` is an alias, you don't know what it will end up resolving to, and copying those unknown types can potentially be really expensive.
the same change should be made in the explicit `Numeric<double>` class definition.
- [ ] revise